### PR TITLE
Testing import comments process on non-generic collections

### PR DESCRIPTION
**Pull request by cartermp**
PR'd on top of my fork so we don't create noise on the main repo.

Here's what the output looks like for a small assembly like `System.Collections.NonGeneric`.  As you can see, there's some work that'll be needed:
- Auditing for changes to method signatures
- Merging the inserted `///` comments with the existing `//` comments to make sure the most up-to-date comments are displayed
- Potential cleanup/reformatting of `///` comments depending on how we want them to be presented in the source

/cc @joshfree @Priya91 @stephentoub @mairaw

### CODE REVIEW

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs [hunk 1]
```diff
@@ -31,6 +31,8 @@ namespace System.Collections
 #if FEATURE_NETCORE
     [FriendAccessAllowed]
 #endif
+    /// <summary>Implements the <see cref="T:System.Collections.IList" /> interface using an array whose size is dynamically increased as required.To browse the .NET Framework source code for this type, see the Reference Source.</summary>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: Can we change the tool to have a max length on the content, such that beyond that it uses multiple lines and wraps, e.g. this should be more like:

``` C#
/// <summary>
/// Implements the <see cref="T:System.Collections.IList" /> interface using an array whose size is 
/// dynamically increased as required.To browse the .NET Framework source code for this type, see the 
/// Reference Source.
/// </summary>
```

Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: Also, why do we need that last sentence about browsing the Reference Source?  Is that actually in the MSDN docs, or was that added automatically by the tool?  Seems quite odd, given that a) you're in the source, and b) a tool that uses this to produce docs can add such a message automatically when producing the docs... it shouldn't be in the source.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs [hunk 2]
```diff
@@ -55,6 +57,7 @@ internal ArrayList(bool trash)
         // Constructs a ArrayList. The list is initially empty and has a capacity
         // of zero. Upon adding the first element to the list the capacity is
         // increased to _defaultCapacity, and then increased in multiples of two as required.
+        /// <summary>Initializes a new instance of the <see cref="T:System.Collections.ArrayList" /> class that is empty and has the default initial capacity.</summary>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:60]: As you noted in your PR comments, all of the `//` and `///` comments will need to be rationalized before they're merged.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs [hunk 3]
```diff
@@ -101,6 +110,11 @@ public ArrayList(ICollection c)
         // the internal array used to hold items.  When set, the internal 
         // array of the list is reallocated to the given capacity.
         // 
+        /// <summary>Gets or sets the number of elements that the <see cref="T:System.Collections.ArrayList" /> can contain.</summary>
+        /// <returns>The number of elements that the <see cref="T:System.Collections.ArrayList" /> can contain.</returns>
+        /// <exception cref="T:System.ArgumentOutOfRangeException"><see cref="P:System.Collections.ArrayList.Capacity" /> is set to a value that is less than <see cref="P:System.Collections.ArrayList.Count" />.</exception>
+        /// <exception cref="T:System.OutOfMemoryException">There is not enough memory available on the system.</exception>
+        /// <filterpriority>1</filterpriority>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: What's the purpose of the filterpriority here?  This relates to VB and its EditorBrowsable support, right?  Is it necessary here?

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs [hunk 4]
```diff
@@ -64,6 +67,9 @@ public ArrayList()
         // initially empty, but will have room for the given number of elements
         // before any reallocations are required.
         // 
+        /// <summary>Initializes a new instance of the <see cref="T:System.Collections.ArrayList" /> class that is empty and has the specified initial capacity.</summary>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: We're in ArrayList here... we shouldn't need to spell out the full System.Collections.ArrayList name... can we change the tool to do that?  I'd expect Roslyn's auto-simplify support would help here.  Basically we should use the shortest names possible based on what namespaces/etc. are imported.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs [hunk 5]
```diff
@@ -602,6 +775,11 @@ public virtual int LastIndexOf(Object value, int startIndex, int count)
 #if FEATURE_NETCORE
         [FriendAccessAllowed]
 #endif
+        /// <summary>Returns a read-only <see cref="T:System.Collections.IList" /> wrapper.</summary>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:778]: These comments should be above attributes (even when conditionally defined), rather than below.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: And, what about the Remarks section of the docs?  Should that be brought in as remarks here?  And what's the plan for other aspects of the docs, e.g. the "Thread Safety" section?

File: src/System.Collections.NonGeneric/src/System/Collections/CaseInsensitiveComparer.cs [hunk 6]
```diff
@@ -36,6 +42,10 @@ public CaseInsensitiveComparer(CultureInfo culture)
             _compareInfo = culture.CompareInfo;
         }
 
+        /// <summary>Gets an instance of <see cref="T:System.Collections.CaseInsensitiveComparer" /> that is associated with the <see cref="P:System.Threading.Thread.CurrentCulture" /> of the current thread and that is always available.</summary>
+        /// <returns>An instance of <see cref="T:System.Collections.CaseInsensitiveComparer" /> that is associated with the <see cref="P:System.Threading.Thread.CurrentCulture" /> of the current thread.</returns>
+        /// <filterpriority>1</filterpriority>
+        /// <PermissionSet><IPermission class="System.Security.Permissions.SecurityPermission, mscorlib, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Flags="UnmanagedCode" />      </PermissionSet>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/CaseInsensitiveComparer.cs:48]: Permissions aren't relevant in .NET Core.

File: src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs [hunk 7]
```diff
@@ -16,14 +16,23 @@
...
 {
+    /// <summary>Compares two objects for equivalence, where string comparisons are case-sensitive.</summary>
+    /// <filterpriority>2</filterpriority>
     public sealed class Comparer : IComparer
     {
         private CompareInfo _compareInfo;
+        /// <summary>Represents an instance of <see cref="T:System.Collections.Comparer" /> that is associated with the <see cref="P:System.Threading.Thread.CurrentCulture" /> of the current thread. This field is read-only.</summary>
+        /// <filterpriority>1</filterpriority>
         public static readonly Comparer Default = new Comparer(CultureInfo.CurrentCulture);
+        /// <summary>Represents an instance of <see cref="T:System.Collections.Comparer" /> that is associated with <see cref="P:System.Globalization.CultureInfo.InvariantCulture" />. This field is read-only.</summary>
```
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs:27]: This may just be a personal preference, but I like having multiple fields without blank lines between them, e.g.

``` C#
private CompareInfo _compareInfo;
public static readonly Comparer Default = new Comparer(CultureInfo.CurrentCulture);
```

but if there are XML comments on the fields, I think there should be a blank line before the comment, e.g.

``` C#
private CompareInfo _compareInfo;

/// <summary>...</summary>
public static readonly Comparer Default = new Comparer(CultureInfo.CurrentCulture);
```

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 2 above)
Comment (joshfree) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:60]: @cartermp thanks for chatting this morning; as discussed let's start by finalizing the dev guidance docs (rationalizing // vs ///, line lengths, etc) then start running the tool to update projects - work with @priya91 for any help needed here;  lastly, please think about what test cases to add to validate intellisense comments (smart quotes, missing comments on new classes, etc).  Thanks

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 1 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: > Can we change the tool to have a max length on the content, such that beyond that it uses multiple lines and wraps

@Priya91, can you comment here?  We use the Roslyn code formatter, so if we can configure it to do this, then yes.

> Also, why do we need that last sentence about browsing the Reference Source?

This is in the intellisense files, so we'll have to remove it either in those files, or during the PR process on a per-file basis.

> And, what about the Remarks section of the docs? Should that be brought in as remarks here? And what's the plan for other aspects of the docs, e.g. the "Thread Safety" section?

@mairaw could clarify here, but I don't believe `<remarks>...</remarks>` comments are generated when building intellisense files.  I can't find any remarks or Thread Safety sections in the intellisense files I've been working with.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 3 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: No, definitely not necessary.  We'll have to remove these before anything gets checked in (either in the intellisense files themselves or on source files the tool is run on).

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 4 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: @Priya91 or @jaredpar could you clarify here?

I believe we'll need to full ID to be able to `f1` the type and open reference docs, but if there's a way to simplify what's shown in the source and still keep that functionality we should definitely do that.  I agree, the fully-qualified ID is too wordy for the source.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 5 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:778]: Good point, we'll look into this.  Running it on other files it looks like the conditionally defined attributes are the only ones it inserts below.

File: src/System.Collections.NonGeneric/src/System/Collections/CaseInsensitiveComparer.cs (hunk 6 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/CaseInsensitiveComparer.cs:48]: Yup, we'll have to remove these as with the filterpriority sections.

File: src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs (hunk 7 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs:27]: Yeah, I prefer that as well.  Is there a guideline on spacing that we can try to get the formatter to adhere to?

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 1 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: @stephentoub @cartermp we don't want remarks and other parts of the documentation here. First, they are not part of the IntelliSense files. Second, the plan is to add that kind of content via Markdown files.

That last sentence about browsing is actually in the summary for a lot of the APIs. Let me try to change the tool to prevent that from making it into the source.

For formatting the comments, I can investigate if it can be done before the code formatter too. What would be the desirable max length here? And do you only want to break into multiple lines if it goes past that length or do you always want to do that?

Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: Thanks.

> For formatting the comments, I can investigate if it can be done before the code formatter too. What would be the desirable max length here?

Something around 80 or 100 is probably reasonable.

>  And do you only want to break into multiple lines if it goes past that length or do you always want to do that?

My preference would be only if it goes past that length.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 2 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:60]: We also need to decide whether we want to keep existing /// comments or replace them with the content coming from IntelliSense. Currently, the tool doesn't replace the /// comments, if they already exist.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 4 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: Isn't the full ID needed for the IntelliSense files to work?

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 3 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: I can also change the code to remove those. But that means IntelliSense won't have that either (not sure if it's necessary, I just know it's there today). I'm not familiar with this and how this was generated in the past (our current tool doesn't add those and that could be a bug)

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 5 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:778]: Let me look into that.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: No. The compiler will expand out whatever is necessary.  It's just like references in code: you only need to provide enough of the name so the compiler understands what you're referring to, but in the generated MSIL it uses the full name.

For example, if I write (assuming I have a `using System.Runtime.CompilerServices;` import in my file):

``` C#
/// <summary>
/// A reference to <see cref="AsyncTaskMethodBuilder"/>.
/// </summary>
```

the compiler generates the following into the doc file:

``` XML
<summary>
A reference to <see cref="T:System.Runtime.CompilerServices.AsyncTaskMethodBuilder"/>.
</summary>
```

And all IntelliSense, refactorings, etc. work as you'd expect in the code.

Comment (stephentoub) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: And you can see that Roslyn offers to simplify the name, both in graying out the namespace and with a quick fix:
image

Comment (jaredpar) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: I just verified the simplifier should properly reduce cref names.  

I **believe** that part of the problem here is the rewriter is not creating a representative compilation.  Instead it's essentially creating a one file project with no references.  This is likely causing the Simplifier to not be able to act here because it can't resolve the references.

File: src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs (hunk 7 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs:27]: I can also check if I can programmatically change that.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 2 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:60]: Sounds like a case-by-base basis here.  We could have the tool insert comments after the existing /// comments and decide if we pick one over the other, merge content, etc.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 4 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: I'm still learning Roslyn here so any feedback and advice is welcome.
Do I have to explicitly call the Simplifier? I didn't have any references before, so I've added the reference to mscorlib to solve the CommentID issue I was having before. But I could still be missing tons of other needed references.

Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: @jaredpar does that mean [this](https://github.com/dotnet/import-comments/blob/master/ImportComments/Program.cs#L45) will have to change such that it picks up the full project references?

**Comment by Priya91** [comment]
@cartermp Can you also opensource the tool used here? Just want to look at all the cases handled.. It's easier that way, than to infer new code comment patterns from tool output ..

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 4 above)
Comment (jaredpar) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:70]: @mairaw correct.  You will need to both tag the nodes that you want to be simplified, in this case the `cref`, and then call the `Simplifier` on the tree.  There is an example in code formatter that you should be able to pattern off of:

https://github.com/dotnet/codeformatter/blob/849fcf60f6b6d85a34e28a9a4d42d82b5ed03116/src/Microsoft.DotNet.CodeFormatting/Rules/ExplicitThisRule.cs

@cartermp correct.  That code needs to pick up the correct set of references for the file.  Alternatively you can use the `MSBuildWorkspace` on the appropriate project file and it will populate with the correct references.

**Comment by cartermp** [comment]
@Priya91 You should be able to see it [here](https://github.com/dotnet/import-comments) - let me know if you can't.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 1 above)
Comment (Priya91) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:34]: > @Priya91, can you comment here? We use the Roslyn code formatter, so if we can configure it to do this, then yes.

@cartermp I used the Roslyn Syntax Visualizer tool on Visual studio. It shows the whole 

```
/// <summary>
/// aaaaaaaaaaaaaaaaa
/// bbbbbbbbbbbbbbbbb
/// </summary>
```

as a `SingleLineDocumentationCommentTrivia`, and the `/// aaaaaaaaaaa` part as `XmlTest` with the following child nodes, `XmlTextLiteralToken` and `XmlTextLiteralNewLineToken`

So you would have to create a `XmlText` node with `XmlTextLiteralNewLineToken` inserted after each `XmlTextLiteralToken`, and the number of `XmlTextLiteralToken` will be equal to the number of n char blocks, where n is the length of line in chars. You also need to figure out a way to not cut words, that is split the string at whitespace chars. Also there is the issue of finding the length, if you are merging with existing documentation comment, then you may have to figure out to use original comment length or rewrite the original comment with the new set length.

I don't think there is a single Roslyn API that will do this for you. @srivatsn Is there an API that will smartly split a long documentation comment into multiple lines? If not, the method outlined above correct?

**Comment by Priya91** [comment]
@cartermp Can you add this as a new project to codeformatter like [DeadRegions](https://github.com/dotnet/codeformatter/tree/master/src/DeadRegions)? So it's easier to use in dev workflow.. Also can you add some tests for the testcases covered using the tool, you can follow the examples in codeformatter repo. 

Some scenario questions:
1. What's the plan on merging with existing // and /\* */ code comments, how to figure out if those comments are doc comments or general note to logic used in code.. 
2. What are the whitespace rules with these doc comments.. 
3. General formatting check should be implemented, after inserting the doc comments.
4. Are the doc comments inserted over every public member and type? If there are no comments after the tool is done, are we reporting the cases where there are no comments.
5. Basic warning checking against a list of allowed and not-allowed xml elements, not having any empty comments, etc
6. Merging with existing doc comments
   - which copy wins? intellisense or the code, or whichever is latest, to ensure recent dev comments are not overridden..
   - say if the existing xml comment has `<para>`, are we nesting this `<para>` inside `<summary>` or `<remarks>` or `<returns>` tag? Even though `<para>` outside these other tags doesn't cause error, it is intended to be used inside another meaningful xml tag for text structure. There may be other elements for which we need to define a general rule.
   - What if the intellisense doc file, has a syntax like this:

```
<summary>
      this is a para
      <returns>some text here</returns>
</summary>
```

how to merge this with a type that has the duplicate definitions, under different nesting :

```
/// <returns>something is returned</returns>
public int mymethod() { return 0; }
```

File: src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs (hunk 7 above)
Comment (Priya91) [inline src/System.Collections.NonGeneric/src/System/Collections/Comparer.cs:27]: Yes, just add a `EndOfLineTrivia` token as the first lead trivia of the node..

**Comment by mairaw** [comment]
Answering a few of your questions:
1. The tool leaves those untouched. So we need to manually remove those if they are just repeating the same information that will be in the /// comments.
2 and 3. I'm not sure if I understand your point here. Can you provide more info? 
4. For every public API (members and types) that exist in the current .NET Framework IntelliSense. So, for example, if there is an API that only exists in .NET Core, we won't get doc comments for those now. We are supposed to have a tool later to tell us which APIs are undocumented so those can be fixed at a later time too.
5. This tool is only meant to import the content from our existing IntelliSense files as /// comments into the source. So not sure why there's a need to validate the xml elements, empty comments, since the input is not just any source since those are the files we already shipped with VS and was meant to be run only once. If we're thinking about a broader scope and usage, then we'd need to add some checking, I agree.
6. About merging:
- Currently the tool is leaving the existing /// comments and other kinds of comments untouched. But after discussing this with @cartermp, he actually prefers that I add the content from the IntelliSense too and we manually choose which one to keep on a case-by-case basis. IntelliSense has no timestamp so it's hard to programmatically identify which one is newer.
- The content coming from IntelliSense doesn't have `<para>` tags or `<remarks>`
- The IntelliSense doc file doesn't have such a syntax and we'll be cleaning up merges manually

Please let me know if you have questions.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 5 above)
Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:778]: This is happening because if the Attribute is inside a conditional markup, it is considered part of the leading trivia, different than what happens when it's a regular attribute. I'll probably have to test if inside the leading trivia, there's conditional markup and then insert the comment at a different position instead of just after the last leading trivia.

**Comment by cartermp** [comment]
@Priya91 we're also going to be creating a doc for `///` comment guidelines for corefx.  We'll include a section on non-`///` comments and how they fit in.  It'll drive any changes we need to make for the import process, and future `///` comments.

We'll also eventually have some automated tests that validate intellisense comments (smart quotes, missing `///` comments on public APIs, etc).  That's more of a long-term thing though.

File: src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs (hunk 3 above)
Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: /cc @AnthonyDGreen

Comment (cartermp) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: Talked with @AnthonyDGreen and we can remove these.

Comment (mairaw) [inline src/System.Collections.NonGeneric/src/System/Collections/ArrayList.cs:117]: OK, good to know.
